contributors/design-proposals/kubeadm-upgrades.md

```diff
@@ -0,0 +1,452 @@
+
+* gcr.io/google_containers/kube-proxy: 5/5 pulled
+
+Upgrading the API Server DaemonSet... 
+
+* Applying the new version of the DaemonSet... ✓
+
+* Verifying that it comes up cleanly... ✓
+
+Upgrading the Controller-manager DaemonSet...
+
+* Applying the new version of the DaemonSet... ✓
+
+* Verifying that it comes up cleanly... ✓
+
+Upgrading the Scheduler DaemonSet...
+
+* Applying the new version of the DaemonSet... ✓
+
+* Verifying that it comes up cleanly... ✓
+
+Upgrading the kube-proxy DaemonSet... ✓
+
+Upgrading the kube-dns Deployment... ✓
+
+Success! Your cluster was upgraded to v1.8.3
+
+UX when upgrading a minor version:
+
+$ kubeadm upgrade plan
+
+Inspecting current cluster state:
+
+ -> Self-hosting active ✓
+
+ -> Configuration exists in-cluster ✓
+
+ --> Validating configuration ✓
+
+ -> Fetching cluster versions ✓
+
+ --> Control plane version: v1.7.5
+
+ --> Latest stable patch version: v1.7.8
+
+ --> Latest stable minor version: v1.8.3
+
+ --> Kubeadm CLI version: v1.8.1
+
+ ---> Note! The kubeadm CLI should be upgraded
+      to v1.8.3 if you want to upgrade your
+      Cluster to v1.8.3
+
+ Both minor and patch release upgrades available!
+
+ Upgrade to the latest patch version of your branch:
+
+			Current version	Upgrade available
+
+ API server:		v1.7.5		v1.7.8
+
+ Controller-manager:	v1.7.5		v1.7.8
+
+ Scheduler:		v1.7.5		v1.7.8
+
+ Kube-proxy:		v1.7.5		v1.7.8
+
+ Kube-dns:			v1.14.2		v1.14.4
+
+ Etcd:			v3.0.17		v3.1.10
+
+ You can now apply the upgrade by executing the following command:
+
+	kubeadm upgrade apply --version v1.7.8
+
+Upgrade to the latest stable minor release:
+
+			Current version	Upgrade available
+
+ API server:		v1.7.5		v1.8.3
+
+ Controller-manager:	v1.7.5		v1.8.3
+
+ Scheduler:		v1.7.5		v1.8.3
+
+ Kube-proxy:		v1.7.5		v1.8.3
+
+ Kube-dns:			v1.14.2		v1.15.5
+
+ You can now apply the upgrade by executing the following command:
+
+	kubeadm upgrade apply --version v1.8.3
+
+ Components that must be upgraded manually:
+
+ Kubelet			2 x v1.8.0		v1.8.3
+
+				3 x v1.8.1		v1.8.3
+
+# Implementation options of the upgrade
+
+High-level approaches:
+
+1. Build the upgrade logic into kubeadm
+
+2. Build an upgrading Operator that does the upgrading for us
+
+    * Would consume a TPR with details how to do the upgrade
+
+The historical preference was to build a reusable Operator that does the upgrading for us. *Either way we would deliver the same 'kubeadm upgrade' UX.* In the latter case, that would involve kubeadm deploying an upgrading controller on behalf of the user, and tracking its lifecycle through the kubernetes API.
+
+There are pros and cons for each method:
+
+1. Build it into the `kubeadm` CLI command
+
+    1. Pros:
+
+        1. *Simplicity*; Very straightforward to implement; basically doing the same steps as `kubeadm init` would, only slightly modified to fit the upgrading purpose. We will construct a `kubeadm upgrade` command in any case
```

Comment:
You should make these bullets instead of numbers so that the alphabet doesn't keep increasing across top level items.

Reply:
fixed